File: packages/widget/test/session-picker-mount.test.ts
// @vitest-environment jsdom
//
// T11 built the two picker surfaces (session-picker.ts's pre-chat screen and
// header switcher); the screens rewrite retired both in favour of the Home
// and Messages screens (ui/home-screen.ts, ui/messages-screen.ts) plus the
// new-conversation surface (ui/new-conversation.ts). This file proves the
// SAME invariants that mattered before still hold under the new mounting:
// past sessions render, a terminal one stays pickable, picking one joins it,
// and starting a new one mints rather than joins.
//
// ── There is no more "gate that hides a whole surface" ────────────────────
//
// Home and Messages are always mounted and always reachable via the nav
// tabs, regardless of session count — `sessions.length > 0` no longer
// decides whether a SURFACE shows, only whether a ROW does (Home's single
// "Recent conversation" row, Messages' full list). Every test below reflects
// that: "the gate" is now about row presence, not screen visibility.
//
// ── The two actions are not interchangeable ───────────────────────────────
//
// `switchSession` is `session.join` under the hood and deliberately does NOT
// mint a session; `startNewSession` does. Wiring "start a new conversation" to
// the former would drop the customer into whichever conversation the server
// picked instead of the fresh one they asked for, which is why there is an
// explicit test below that they stay distinct.

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import { ChatClientConfigError } from '@dhaam-ccrm/js';

import { mount, unmount } from '../src/index.js';
import type { ChatWidget } from '../src/widget.js';
import type { WidgetConfig } from '../src/config.js';

const PK_TEST = 'dhp_' + 'test_' + '0123456789abcdefghijklmn';

let frameCounter = 0;
const frameId = (): string => `01ARZ3NDEKTSV4RRFFQ69G5F${String(frameCounter++ % 10)}0`;

class FakeWebSocket {
  static instances: FakeWebSocket[] = [];
  readonly sent: string[] = [];
  onopen: (() => void) | null = null;
  onmessage: ((event: { data: unknown }) => void) | null = null;
  onerror: (() => void) | null = null;
  onclose: ((event: { code: number; reason: string; wasClean: boolean }) => void) | null = null;

  constructor(readonly url: string) {
    FakeWebSocket.instances.push(this);
  }
  send(data: string): void {
    this.sent.push(data);
  }
  close(): void {
    this.onclose?.({ code: 1000, reason: '', wasClean: true });
  }
  open(): void {
    this.onopen?.();
  }

  frames(type: string): Array<{ id: string; d: Record<string, unknown> }> {
    return this.sent
      .map((raw) => JSON.parse(raw) as { t: string; id: string; d: Record<string, unknown> })
      .filter((frame) => frame.t === type);
  }

  push(t: string, d: unknown): void {
    this.onmessage?.({ data: JSON.stringify({ v: 1, t, id: frameId(), ts: Date.now(), d }) });
  }

  ack(sessionId = 'sess_current', status = 'ASSIGNED'): void {
    this.push('connection.ack', {
      protocolVersion: 1,
      seq: 0,
      session: {
        sessionId,
        status,
        mode: 'HUMAN',
        participants: [{ participantId: 'cus_1', type: 'CUSTOMER' }],
        createdAt: new Date().toISOString(),
      },
    });
  }
}

function summaryRow(overrides: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    id: 'sess_past',
    status: 'RESOLVED',
    mode: 'HUMAN',
    createdAt: '2026-08-19T09:00:00.000Z',
    closedAt: '2026-08-19T10:00:00.000Z',
    lastMessageAt: '2026-08-19T09:30:00.000Z',
    lastMessagePreview: 'Thanks!',
    unreadCount: 0,
    ...overrides,
  };
}

function config(overrides: Partial<WidgetConfig> = {}): WidgetConfig {
  return {
    auth: { publishableKey: PK_TEST, tokenEndpoint: '/api/chat-token' },
    identity: { userId: 'cus_1' },
    apiUrl: 'https://chat.example.com',
    wsUrl: 'wss://chat.example.com',
    onError: () => undefined,
    ...overrides,
  };
}

let sessionRows: unknown[] = [];

function shadow(): ShadowRoot {
  const element = document.querySelector<HTMLElement>('dh-chat-widget');
  if (element === null) throw new Error('widget host not found');
  const root = element.shadowRoot;
  if (root === null) throw new Error('shadow root not found');
  return root;
}

const query = <T extends Element>(selector: string): T => {
  const found = shadow().querySelector<T>(selector);
  if (found === null) throw new Error(`not found: ${selector}`);
  return found;
};

/** Unlike `query`, returns `null` instead of throwing — for asserting a selector is genuinely ABSENT. */
const find = <T extends Element>(selector: string): T | null => shadow().querySelector<T>(selector);

const visible = (node: HTMLElement): boolean => !node.hidden && node.style.display !== 'none';
/** Whether `node` is on screen INCLUDING its ancestors — `hidden` on a parent (e.g. Home's recent-conversation section) hides everything inside it too. */
const reallyVisible = (node: Element): boolean => {
  let current: Element | null = node;
  while (current !== null) {
    if (current instanceof HTMLElement && !visible(current)) return false;
    current = current.parentElement;
  }
  return true;
};
const homePane = (): HTMLElement => query<HTMLElement>('.dh-home');
/** Home always BUILDS its recent-conversation row; `update()` hides the SECTION around it rather than removing it — see home-screen.ts. So presence alone proves nothing; visibility does. */
const homeShowsRecentRow = (): boolean => {
  const row = homePane().querySelector('.dh-home-recent-row');
  return row !== null && reallyVisible(row);
};
const messagesPane = (): HTMLElement => query<HTMLElement>('.dh-messages');
const messagesRows = (): HTMLButtonElement[] => [
  ...messagesPane().querySelectorAll<HTMLButtonElement>('.dh-messages-row'),
];

/** The nav bar has exactly two tabs; found by label rather than position so a reorder cannot silently break this. */
function navTab(label: 'Home' | 'Messages'): HTMLButtonElement {
  const tab = [...shadow().querySelectorAll<HTMLButtonElement>('.dh-nav-tab')].find(
    (candidate) => candidate.querySelector('.dh-nav-label')?.textContent === label,
  );
  if (tab === undefined) throw new Error(`${label} tab not found`);
  return tab;
}

async function settle(): Promise<void> {
  for (let i = 0; i < 6; i += 1) await new Promise((resolve) => setTimeout(resolve, 0));
}

async function goToMessages(): Promise<void> {
  navTab('Messages').click();
  await settle();
}

/** Sockets whose `connection.hello` has already been answered. */
let handshaked = new WeakSet<FakeWebSocket>();

/**
 * Completes the handshake on every socket the client opened but nobody drove,
 * and returns the newest one.
 *
 * Picking a row calls `switchSession()`, which is a TEARDOWN and a
 * re-establish (see core's `switchToSession`): it drops the socket the widget
 * booted on and opens another, whose `connection.ack` the server resolves on
 * its own — `connection.hello` carries no session id. The explicit
 * `session.join` that follows is what corrects it.
 */
async function driveHandshakes(resolved = 'sess_current'): Promise<FakeWebSocket> {
  for (const instance of FakeWebSocket.instances) {
    if (handshaked.has(instance)) continue;
    handshaked.add(instance);
    instance.open();
    instance.ack(resolved);
    await settle();
  }
  const socket = FakeWebSocket.instances[FakeWebSocket.instances.length - 1];
  if (socket === undefined) throw new Error('no socket was opened');
  return socket;
}

async function openedWidget(
  overrides: Partial<WidgetConfig> = {},
): Promise<{ widget: ChatWidget; socket: FakeWebSocket }> {
  const widget = mount(config(overrides));
  await settle();
  const socket = await driveHandshakes();
  widget.open();
  await settle();
  return { widget, socket };
}

beforeEach(() => {
  localStorage.clear();
  frameCounter = 0;
  sessionRows = [];
  FakeWebSocket.instances = [];
  handshaked = new WeakSet<FakeWebSocket>();
  vi.stubGlobal('WebSocket', FakeWebSocket);
  vi.stubGlobal(
    'fetch',
    vi.fn(async (input: RequestInfo | URL) => {
      const url = String(input);
      if (url.includes('/api/chat-token')) {
        return new Response(JSON.stringify({ accessToken: 'tok', expiresIn: 3600 }), {
          status: 200,
          headers: { 'content-type': 'application/json' },
        });
      }
      if (url.includes('/chat/sessions/customer')) {
        return new Response(JSON.stringify({ success: true, data: { sessions: sessionRows } }), {
          status: 200,
          headers: { 'content-type': 'application/json' },
        });
      }
      return new Response(JSON.stringify({ success: true, data: { messages: [], hasMore: false } }), {
        status: 200,
        headers: { 'content-type': 'application/json' },
      });
    }),
  );
  document.body.innerHTML = '';
});

afterEach(() => {
  unmount();
  vi.unstubAllGlobals();
});

describe('the gate is exactly sessions.length > 0 — now expressed as rows, not surfaces', () => {
  it('Home and Messages both mount for an empty page, with no rows — what a guest gets', async () => {
    sessionRows = [];
    await openedWidget();

    // The launcher opens onto Home either way now — the "conversation is
    // shown instead" behavior this test used to prove no longer applies,
    // since there is no conversation to show yet and nothing was asked for.
    expect(visible(homePane())).toBe(true);
    expect(homeShowsRecentRow()).toBe(false);

    await goToMessages();
    expect(messagesRows()).toHaveLength(0);
    expect(messagesPane().querySelector('.dh-messages-empty')?.textContent).toBe('No conversations yet.');
  });

  it('Home shows the most recent session as its own row', async () => {
    sessionRows = [summaryRow({ id: 'sess_past' })];
    await openedWidget();

    expect(visible(homePane())).toBe(true);
    expect(homeShowsRecentRow()).toBe(true);
  });

  it('Messages renders one row per session', async () => {
    sessionRows = [summaryRow({ id: 'a' }), summaryRow({ id: 'b' })];
    await openedWidget();
    await goToMessages();

    expect(messagesRows()).toHaveLength(2);
  });

  it('keeps a terminal session pickable — reactivation is a real path back', async () => {
    sessionRows = [summaryRow({ id: 'closed_one', status: 'CLOSED' })];
    await openedWidget();
    await goToMessages();

    const row = messagesRows()[0];
    if (row === undefined) throw new Error('no row rendered');
    expect(row.disabled).toBe(false);
    expect(row.closest('.dh-messages-item')?.getAttribute('data-status')).toBe('CLOSED');
  });
});

describe('picking a conversation', () => {
  it('joins the chosen session and puts the conversation back on screen', async () => {
    sessionRows = [summaryRow({ id: 'sess_past' })];
    const { socket } = await openedWidget();
    await goToMessages();

    const row = messagesRows()[0];
    if (row === undefined) throw new Error('no row rendered');
    row.click();
    await settle();

    // The socket the switch opened, not the one the widget booted on: the
    // teardown drops that one before any join is written.
    const live = await driveHandshakes();
    expect(live).not.toBe(socket);
    const joins = live.frames('session.join');
    expect(joins).toHaveLength(1);
    expect(joins[0]?.d['sessionId']).toBe('sess_past');

    // Picking a row is `go('conversation')` — Messages is left behind, not
    // stacked under it.
    expect(visible(messagesPane())).toBe(false);
    expect(visible(query<HTMLElement>('.dh-log'))).toBe(true);
    expect(query<HTMLElement>('.dh-composer').hidden).toBe(false);
  });

  it('does not optimistically flip a terminal session to open', async () => {
    sessionRows = [summaryRow({ id: 'sess_past', status: 'CLOSED' })];
    const { widget } = await openedWidget();
    await goToMessages();

    const before = widget.store.getState().session;
    const row = messagesRows()[0];
    if (row === undefined) throw new Error('no row rendered');
    row.click();
    await settle();

    // The server reactivates on the customer's next message, behind a flag
    // that defaults to OFF. Guessing locally would show a live conversation
    // on a deployment where reactivation never happens.
    expect(widget.store.getState().session).toBe(before);
  });

  it('marks the current conversation in Messages', async () => {
    sessionRows = [summaryRow({ id: 'sess_current', status: 'ASSIGNED', closedAt: null }), summaryRow({ id: 'other' })];
    await openedWidget();
    await goToMessages();

    const current = messagesRows().find((row) => row.getAttribute('aria-current') === 'true');
    expect(current).toBeDefined();
    expect(current?.getAttribute('aria-label')).toContain('current conversation');
  });
});

describe('starting a new conversation', () => {
  /** Opens the new-conversation surface from Messages and fills in the one required field. */
  async function reachStartSurface(): Promise<void> {
    await goToMessages();
    query<HTMLButtonElement>('.dh-messages-new').click();
    await settle();
    query<HTMLTextAreaElement>('.dh-newconvo-message').value = 'Hello';
  }

  it('mints a session rather than joining one', async () => {
    sessionRows = [summaryRow()];
    const { widget, socket } = await openedWidget();

    const startNew = vi.spyOn(widget.store.client, 'startNewSession').mockResolvedValue();
    const switchTo = vi.spyOn(widget.store.client, 'switchSession');

    await reachStartSurface();
    query<HTMLButtonElement>('.dh-newconvo-form .dh-form-submit').click();
    await settle();

    expect(startNew).toHaveBeenCalledTimes(1);
    // The distinction that matters: `switchSession` is `session.join` and
    // deliberately does not mint a session.
    expect(switchTo).not.toHaveBeenCalled();
    expect(socket.frames('session.join')).toHaveLength(0);
  });

  it('leaves the new-conversation surface for the conversation once the session lands', async () => {
    sessionRows = [summaryRow()];
    const { widget } = await openedWidget();
    vi.spyOn(widget.store.client, 'startNewSession').mockResolvedValue();

    await reachStartSurface();
    query<HTMLButtonElement>('.dh-newconvo-form .dh-form-submit').click();
    await settle();

    // `query` throws rather than returning null, so absence is asserted with
    // `find` instead — see this file's own note on the two helpers.
    expect(find('.dh-newconvo-form')).toBeNull();
    expect(query<HTMLElement>('.dh-composer').hidden).toBe(false);
  });

  it('goes busy on the Start button so one round trip cannot mint two sessions', async () => {
    sessionRows = [summaryRow()];
    const { widget } = await openedWidget();

    let release = (): void => undefined;
    vi.spyOn(widget.store.client, 'startNewSession').mockReturnValue(
      new Promise<void>((resolve) => {
        release = resolve;
      }),
    );

    await reachStartSurface();
    const start = query<HTMLButtonElement>('.dh-newconvo-form .dh-form-submit');
    start.click();
    await settle();

    expect(start.disabled).toBe(true);

    // Once `startNewSession` resolves, `startNewConversation` also sends the
    // typed message and closes the surface — so the button that was "busy"
    // is gone by now, replaced by the live conversation. What matters is
    // that exactly one mint happened and the composer is what is left.
    release();
    await settle();
    expect(find('.dh-newconvo-form')).toBeNull();
    expect(query<HTMLElement>('.dh-composer').hidden).toBe(false);
    expect(widget.store.client.startNewSession).toHaveBeenCalledTimes(1);
  });

  it('stays on the surface when the new session could not be opened', async () => {
    sessionRows = [summaryRow()];
    const { widget } = await openedWidget();
    vi.spyOn(widget.store.client, 'startNewSession').mockRejectedValue(new Error('nope'));

    await reachStartSurface();
    query<HTMLButtonElement>('.dh-newconvo-form .dh-form-submit').click();
    await settle();

    // Dropping the customer onto an empty transcript would hide the fact that
    // nothing actually happened, and the typed message would be lost with it.
    expect(find('.dh-newconvo-form')).not.toBeNull();
    expect(query<HTMLTextAreaElement>('.dh-newconvo-message').value).toBe('Hello');
  });
});

describe('the SWITCHED-close pairing', () => {
  it('does not close the conversation the customer just switched INTO', async () => {
    sessionRows = [summaryRow({ id: 'sess_past', status: 'CLOSED' })];
    const { widget, socket } = await openedWidget();
    await goToMessages();

    const row = messagesRows()[0];
    if (row === undefined) throw new Error('no row rendered');
    row.click();
    await settle();

    const live = await driveHandshakes();
    const join = live.frames('session.join')[0];
    if (join === undefined) throw new Error('no session.join frame was sent');
    live.onmessage?.({
      data: JSON.stringify({ v: 1, t: 'ack', id: frameId(), ref: join.id, ts: Date.now(), d: { ok: true } }),
    });
    await settle();

    // The server acks the switch...
    live.push('session.updated', {
      session: {
        sessionId: 'sess_past',
        status: 'WAITING_FOR_AGENT',
        mode: 'HUMAN',
        participants: [{ participantId: 'cus_1', type: 'CUSTOMER' }],
        createdAt: '2026-08-19T09:00:00.000Z',
      },
    });
    await settle();

    // ...and, in either order, closes the OTHER session the customer left
    // behind. One customer, one live conversation.
    live.push('session.closed', { sessionId: 'sess_current', closeReason: 'SWITCHED' });
    await settle();

    // The session just switched into must survive both halves of that pair.
    expect(widget.store.getState().session?.id).toBe('sess_past');
    expect(query<HTMLElement>('.dh-system').hidden).toBe(true);
    expect(query<HTMLTextAreaElement>('.dh-input').disabled).toBe(false);
  });
});

describe('an embed whose client has no session summary source', () => {
  it('mounts, connects and chats exactly as before — with no rows and no error', async () => {
    const errors: unknown[] = [];
    const widget = mount(config({ onError: (error) => errors.push(error) }));
    await settle();
    const socket = FakeWebSocket.instances[0];
    if (socket === undefined) throw new Error('no socket');
    socket.open();
    socket.ack();
    await settle();

    vi.spyOn(widget.store.client, 'listSessions').mockRejectedValue(
      new ChatClientConfigError('listSessions() requires config.sessionSummarySource'),
    );

    widget.open();
    await settle();

    expect(homeShowsRecentRow()).toBe(false);
    await goToMessages();
    expect(messagesRows()).toHaveLength(0);

    // A configuration fact, not a fault: reporting it would fire on every page
    // load of an embed that simply has no session list.
    expect(errors).toHaveLength(0);
  });

  it('still reports a genuine failure to load the list', async () => {
    const errors: unknown[] = [];
    const widget = mount(config({ onError: (error) => errors.push(error) }));
    await settle();
    const socket = FakeWebSocket.instances[0];
    if (socket === undefined) throw new Error('no socket');
    socket.open();
    socket.ack();
    await settle();
    errors.length = 0;

    vi.spyOn(widget.store.client, 'listSessions').mockRejectedValue(new Error('503'));

    widget.open();
    await settle();

    expect(errors).toHaveLength(1);
    expect(homeShowsRecentRow()).toBe(false);
  });
});

describe('a host that named the session it wants', () => {
  it('opens directly onto the conversation, with Messages still reachable', async () => {
    sessionRows = [summaryRow()];
    await openedWidget({ sessionId: 'sess_named' });

    // Being pointed at one conversation is not the same as being locked into
    // it: the launcher opens straight onto it (skipping Home, which a host
    // that named a session has effectively overridden), but the nav bar
    // still gets there and Messages still renders the full list.
    expect(query<HTMLElement>('.dh-composer').hidden).toBe(false);
    expect(visible(homePane())).toBe(false);

    await goToMessages();
    expect(messagesRows()).toHaveLength(1);
  });
});

File: packages/flutter/test/ui/session_picker/session_picker_mount_test.dart
// Reproduces the transferable half of `session-picker-mount.test.ts` and
// `session-switch.test.ts` — both picker surfaces wired to a real
// `ChatWidgetCubit`, rather than to spies.
//
// ── What transfers, and what does not ────────────────────────────────────
//
// TRANSFERS: the gate is exactly `sessions.length > 0` and is asked by the
// caller; a terminal row stays pickable; picking JOINS the chosen session
// and puts the conversation back on screen at once; picking does NOT
// optimistically flip a terminal session to open; the current conversation
// is marked.
//
// DOES NOT: everything `session-switch.test.ts` asserts about the rendered
// TRANSCRIPT swapping — page-one being refetched with no cursor from the old
// session, and the choice surviving a reload. Those are core and history
// concerns (`dhaam_chat` plus the REST history adapter), not the picker's,
// and this package's `WidgetChatClient.joinSession` is a frame write with no
// history behaviour behind it. Nor does "reports a refused switch instead of
// leaking an unhandled rejection": `joinSession` returns no future here, so
// there is no rejection to leak.

import 'package:dhaam_chat/dhaam_chat.dart';
import 'package:dhaam_chat_flutter/dhaam_chat_flutter.dart';
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:flutter_test/flutter_test.dart';

import '../../state/fake_widget_chat_client.dart';

ChatSessionSummary _summary({
  String id = 'sess_past',
  ChatStatus status = ChatStatus.resolved,
}) =>
    ChatSessionSummary(
      id: id,
      status: status,
      mode: ChatMode.human,
      createdAt: DateTime.utc(2026, 8, 19, 9),
      lastMessageAt: DateTime.utc(2026, 8, 19, 9, 30),
      lastMessagePreview: 'Thanks!',
    );

void main() {
  late FakeWidgetChatClient client;
  late ChatWidgetCubit cubit;

  setUp(() {
    client = FakeWidgetChatClient();
    cubit = ChatWidgetCubit(client: client);
  });

  tearDown(() async {
    await cubit.close();
    await client.dispose();
  });

  /// Mounts [child] over the live Cubit, rebuilding it on every state — the
  /// way a real mounting layer does, rather than snapshotting the state once
  /// at construction and leaving the test to get its ordering right.
  Widget mount(Widget Function(ChatWidgetState state) child) {
    return MaterialApp(
      home: Scaffold(
        body: BlocProvider<ChatWidgetCubit>.value(
          value: cubit,
          child: BlocBuilder<ChatWidgetCubit, ChatWidgetState>(
            builder: (BuildContext context, ChatWidgetState state) =>
                child(state),
          ),
        ),
      ),
    );
  }

  /// The picker screen, wired to the Cubit exactly as a mounting layer would.
  Widget mountScreen() {
    return mount(
      (ChatWidgetState state) => SessionPickerScreen(
        sessions: state.sessionSummaries,
        onSelect: cubit.selectSession,
        onStartNew: cubit.startNewConversation,
        cornerRadius: chatCornerRadius(state.config),
      ),
    );
  }

  group('the gate is exactly sessions.length > 0, and the caller asks it', () {
    testWidgets('an empty page renders the surface, with no rows — a guest',
        (WidgetTester tester) async {
      await tester.pumpWidget(mountScreen());

      // Not a hidden component: the surface is complete and says so.
      expect(find.byType(SessionPickerScreen), findsOneWidget);
      expect(find.byType(SessionRow), findsNothing);
      expect(find.text(kSessionPickerEmptyText), findsOneWidget);
      expect(find.text(kStartNewConversationLabel), findsOneWidget);
    });

    testWidgets('one row per session', (WidgetTester tester) async {
      cubit.updateSessionSummaries(<ChatSessionSummary>[
        _summary(id: 'a'),
        _summary(id: 'b'),
      ]);
      await tester.pumpWidget(mountScreen());

      expect(find.byType(SessionRow), findsNWidgets(2));
    });

    testWidgets('keeps a terminal session pickable — reactivation is a path',
        (WidgetTester tester) async {
      cubit.updateSessionSummaries(<ChatSessionSummary>[
        _summary(id: 'closed_one', status: ChatStatus.closed),
      ]);
      await tester.pumpWidget(mountScreen());

      expect(find.text('Closed'), findsOneWidget);
      await tester.tap(find.byType(SessionRow));
      await tester.pump();

      expect(client.joinedSessionIds, <String>['closed_one']);
    });
  });

  group('picking a conversation', () {
    testWidgets('joins the chosen session and shows the conversation at once',
        (WidgetTester tester) async {
      cubit.updateSessionSummaries(<ChatSessionSummary>[_summary()]);
      await tester.pumpWidget(mountScreen());

      await tester.tap(find.byType(SessionRow));
      await tester.pump();

      expect(client.joinedSessionIds, <String>['sess_past']);
      // The pane flips straight away rather than on completion: holding the
      // picker up for a round trip would read as a dead tap.
      expect(cubit.state.screen, ScreenName.conversation);
      // The picker IS one of the places the widget deliberately puts a
      // conversation in front of the customer, so the pre-chat gate can arm.
      expect(cubit.state.conversationOpened, isTrue);
    });

    testWidgets('does not optimistically flip a terminal session to open',
        (WidgetTester tester) async {
      cubit.updateSessionSummaries(<ChatSessionSummary>[
        _summary(id: 'sess_past', status: ChatStatus.closed),
      ]);
      await tester.pumpWidget(mountScreen());

      final SessionSnapshot? before = cubit.state.session;
      await tester.tap(find.byType(SessionRow));
      await tester.pump();

      // The server reactivates on the customer's next message, behind a flag
      // that defaults to OFF. Guessing locally would show a live
      // conversation on a deployment where reactivation never happens.
      expect(cubit.state.session, same(before));
    });

    testWidgets('goes through the same funnel a Messages row does',
        (WidgetTester tester) async {
      // `selectSession` is a forwarder to `openConversation`, so a picker row
      // and a Messages row cannot diverge on what picking means. Asserted by
      // driving both and comparing the state each leaves behind.
      cubit.updateSessionSummaries(<ChatSessionSummary>[_summary(id: 'x')]);
      await tester.pumpWidget(mountScreen());

      await tester.tap(find.byType(SessionRow));
      await tester.pump();
      final ChatWidgetState afterPicker = cubit.state;

      cubit.openConversation('x');
      expect(cubit.state, afterPicker);
      expect(client.joinedSessionIds, <String>['x', 'x']);
    });

    testWidgets('discards a surface left open on the way in',
        (WidgetTester tester) async {
      // Asking for a DIFFERENT conversation is one of the two moments that
      // mean "I am done with this". Without the discard the new conversation
      // is drawn UNDER a stale form.
      cubit.updateSessionSummaries(<ChatSessionSummary>[_summary(id: 'x')]);
      cubit.startNewConversation();
      expect(cubit.state.activeSurface, isA<ComposingNewSurface>());

      await tester.pumpWidget(mountScreen());
      await tester.tap(find.byType(SessionRow));
      await tester.pump();

      expect(cubit.state.activeSurface, isNull);
      expect(cubit.state.composingNew, isFalse);
    });
  });

  group('the switcher, mounted in a conversation', () {
    // Centred, not flush against the edge. The popover is anchored to the
    // toggle's right edge and is 300 wide, so a toggle in the top-left
    // corner puts most of its own panel off-screen — where a tap on a row
    // hit-tests nothing. A header places this with room around it; a test
    // that does not is testing a layout no host would ship.
    Widget mountSwitcher() {
      return mount(
        (ChatWidgetState state) => Center(
          child: SessionSwitcher(
            sessions: state.sessionSummaries,
            currentSessionId: state.session?.sessionId,
            onSelect: cubit.selectSession,
            onStartNew: cubit.startNewConversation,
          ),
        ),
      );
    }

    testWidgets('marks the conversation the customer is in, and joins another',
        (WidgetTester tester) async {
      final SemanticsHandle handle = tester.ensureSemantics();
      await tester.pumpWidget(mountSwitcher());

      // The snapshot arrives on a stream, so its listener runs on a
      // microtask. `pump` is what flushes those — an `await Future.delayed`
      // here would hang forever, because `testWidgets` runs the body in a
      // fake-async zone whose clock only advances when the test pumps.
      client.emitSession(testSession(id: 'sess_current'));
      cubit.updateSessionSummaries(<ChatSessionSummary>[
        _summary(id: 'sess_current', status: ChatStatus.assigned),
        _summary(id: 'other'),
      ]);
      await tester.pump();

      await tester.tap(find.byIcon(Icons.list_rounded));
      await tester.pumpAndSettle();

      expect(
        tester
            .getSemantics(find.byKey(const ValueKey<String>('sess_current')))
            .label,
        contains('current conversation'),
      );

      await tester.tap(find.byKey(const ValueKey<String>('other')));
      await tester.pumpAndSettle();

      expect(client.joinedSessionIds, <String>['other']);
      expect(cubit.state.screen, ScreenName.conversation);
      handle.dispose();
    });

    testWidgets('starting fresh from inside mints rather than joins',
        (WidgetTester tester) async {
      // The distinction that matters: joining drops the customer into
      // whichever conversation the server picked, not the fresh one they
      // asked for.
      cubit.updateSessionSummaries(<ChatSessionSummary>[_summary()]);
      await tester.pumpWidget(mountSwitcher());
      await tester.tap(find.byIcon(Icons.list_rounded));
      await tester.pumpAndSettle();

      await tester.tap(find.text(kStartNewConversationLabel));
      await tester.pumpAndSettle();

      expect(client.joinedSessionIds, isEmpty);
      expect(cubit.state.activeSurface, isA<ComposingNewSurface>());
      // A form opened from the switcher is a detour, not a conversation —
      // counting it would arm the pre-chat gate behind the very form that is
      // already asking those questions.
      expect(cubit.state.conversationOpened, isFalse);
    });
  });
}
